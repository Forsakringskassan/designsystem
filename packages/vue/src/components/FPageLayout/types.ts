/**
 * @public
 */
export interface FPageLayoutSlotMapping {
    simple: ["header", "content", "footer"];
    "left-panel": ["header", "left", "content", "footer"];
    "right-panel": ["header", "right", "content", "footer"];
    "three-column": ["header", "left", "right", "content", "footer"];
}

/**
 * @public
 */
export type FPageLayoutType = Exclude<keyof FPageLayoutSlotMapping, never>;

/**
 * @public
 */
export type FPageLayoutBindings<K extends string> = K extends FPageLayoutType
    ? {
          [Property in FPageLayoutSlotMapping[K][number]]: `${Property}`;
      }
    : never;
