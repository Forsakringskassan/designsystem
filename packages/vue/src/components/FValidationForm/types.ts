/**
 * @public
 */
export enum FValidationFormAction {
    CONTINUE = 0,
    CANCEL = 1,
}

/**
 * @public
 */
export type FValidationFormResult =
    | void // eslint-disable-line @typescript-eslint/no-invalid-void-type -- allow implicit undefined return value
    | FValidationFormAction
    | Promise<FValidationFormAction>
    | Promise<void>;

/**
 * @public
 */
export type FValidationFormCallback = () => FValidationFormResult;
