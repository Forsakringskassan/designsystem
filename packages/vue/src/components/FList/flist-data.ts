import { type ListArray, type ListItem } from "../../types";

/**
 * @public
 */
export interface FListData {
    selectedItems: ListArray;
    activeItem: ListItem | undefined;
}
