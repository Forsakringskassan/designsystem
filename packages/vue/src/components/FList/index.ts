export { default as FList } from "./FList.vue";
