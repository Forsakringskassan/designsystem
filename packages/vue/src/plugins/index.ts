export * from "./test";
export * from "./translation";
export * from "./validation";
export * from "./error";
export { FormatPlugin } from "./format";
