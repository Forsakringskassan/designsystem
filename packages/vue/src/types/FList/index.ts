/**
 * @public
 */
export type UnknownItem = Record<string, unknown>;

/**
 * Represents a single item in a list-based component (FList, FTable, etc).
 *
 * @public
 */
export type ListItem<T extends object = UnknownItem> = T;

/**
 * Represents an array of items in a list-based component (FList, FTable, etc).
 *
 * @public
 */
export type ListArray<T extends object = UnknownItem> = Array<ListItem<T>>;
