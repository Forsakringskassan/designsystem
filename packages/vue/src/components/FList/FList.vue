<script setup lang="ts" generic="T extends object">
import { type Ref, computed, getCurrentInstance, onMounted, ref, useSlots, watch, type PropType } from "vue";
import { ElementIdService } from "@fkui/logic";
import { FCheckboxField } from "../FCheckboxField";
import { itemEquals, includeItem, handleKeyboardFocusNavigation, getElementFromVueRef } from "../../utils";
import { getInternalKey, setInternalKeys } from "../../utils/internal-key";
import { useTranslate } from "../../plugins";
import { ActivateItemInjected } from "../FCrudDataset";

const $t = useTranslate();
const slots = useSlots();
const { registerCallbackAfterItemAdd, registerCallbackBeforeItemDelete } = ActivateItemInjected<T>();
const internalKey = getInternalKey<T>();

const selectedItems: Ref<T[]> = ref([]);
const activeItem = ref<T | undefined>(undefined);
const ulElement = ref<HTMLElement | null>();

const props = defineProps({
    /**
     * The items to be listed.
     * The items will be listed in the given array order.
     */
    items: {
        type: Array as PropType<T[]>,
        required: true,
    },
    /**
     * Unique attribute in items.
     * V-for directive in vue requires a unique key to
     * be able to optimize reuse of components.
     */
    keyAttribute: {
        type: String,
        required: false,
        default: undefined,
    },
    /**
     * If `true` the list will be selectable.
     * @see 'select' and 'unselect' events.
     */
    selectable: {
        type: Boolean,
        default: false,
    },
    /**
     * Only applies if selectable:true
     * Disable checkbox in interactive list and create a <a>-wrapper for each item, making them clickable
     * @see 'select' and 'unselect' events.
     */
    checkbox: {
        type: Boolean,
        default: true,
    },
    /**
     * V-model will bind to value containing selected items.
     */
    modelValue: {
        type: Array as PropType<T[] | undefined>,
        required: false,
        default: () => undefined,
    },
    /**
     * V-model will bind to value containing the current active item.
     */
    active: {
        type: Object as PropType<T | undefined>,
        required: false,
        default: () => undefined,
    },
    /**
     * Unique item id that will be used instead of the automatically generated one.
     */
    elementId: {
        type: String,
        default: () => ElementIdService.generateElementId(),
    },
});

const emit = defineEmits<{
    /**
     * Emitted when item is activated, i.e. clicked
     */
    change: [item: T];
    /**
     * Emitted when item is activated, i.e. clicked
     */
    click: [item: T];
    /**
     * Emitted when item is selected.
     */
    select: [item: T];
    /**
     * Emitted when item is unselected.
     */
    unselect: [item: T];
    /**
     * V-model active event.
     */
    "update:active": [item: T];
    /**
     * V-model event to update value property.
     */
    "update:modelValue": [item: T[]];
}>();

const isEmpty = computed((): boolean => {
    return internalItems.value.length === 0;
});

const internalItems = computed((): T[] => {
    const { keyAttribute } = props;
    if (keyAttribute) {
        return setInternalKeys(props.items, keyAttribute as keyof T);
    }

    return setInternalKeys(props.items);
});

watch(
    () => props.items,
    () => {
        updateSelectedItemsFromVModel();
    },
    { deep: true, immediate: true },
);

watch(
    () => props.modelValue,
    () => {
        updateSelectedItemsFromVModel();
    },
    { deep: true, immediate: true },
);

watch(
    () => props.active,
    () => {
        updateActiveItemFromVModel();
    },
    { immediate: true },
);

onMounted(() => {
    if (props.selectable && props.checkbox) {
        if (!slots["screenreader"]) {
            throw Error('Slot "screenreader" is required when having "selectable" & "checkbox" option.');
        }
        registerCallbackAfterItemAdd(callbackAfterItemAdd);
        registerCallbackBeforeItemDelete(callbackBeforeItemDelete);
    }
});

function getLiElements(): HTMLElement[] {
    const element = getElementFromVueRef(ulElement.value);
    return Array.from(element.children) as HTMLElement[];
}

function itemKey(item: T): string {
    return String(item[internalKey]);
}

function isSelected(item: T): boolean {
    return includeItem(item, selectedItems.value, internalKey);
}

function itemClasses(item: T): Record<string, boolean> {
    return {
        "list__item--selected": isSelected(item),
        "list__item--active": isActive(item),
    };
}

function onSelect(item: T): void {
    if (includeItem(item, selectedItems.value, internalKey)) {
        selectedItems.value = selectedItems.value.filter((i) => !itemEquals(i, item, internalKey));
        emit("unselect", item);
    } else {
        selectedItems.value.push(item);
        emit("select", item);
    }

    updateVModelWithSelectedItems();
    getCurrentInstance()?.proxy?.$forceUpdate();
}

function setActiveItem(item: T): void {
    emit("click", item);

    if (!itemEquals(item, activeItem.value, internalKey)) {
        emit("change", item);
        activeItem.value = item;
        emit("update:active", activeItem.value);
    }
}

function onItemClick(event: Event, index: number, item: T): void {
    setActiveItem(item);
}

function updateVModelWithSelectedItems(): void {
    if (props.modelValue) {
        emit("update:modelValue", selectedItems.value);
    }
}

function updateSelectedItemsFromVModel(): void {
    if (Array.isArray(props.modelValue)) {
        // Remove old selected items that may not exists in items.
        selectedItems.value = props.modelValue.filter((item) => {
            return includeItem(item, internalItems.value, internalKey);
        });
    } else {
        selectedItems.value = [];
    }
}

function updateActiveItemFromVModel(): void {
    if (props.active === undefined) {
        activeItem.value = undefined;
    } else if (!itemEquals(props.active, activeItem.value, internalKey)) {
        activeItem.value = props.active;
    }
}

function onItemKeyDown(event: KeyboardEvent, item: T): void {
    switch (event.key) {
        case "Up":
        case "Down":
        case "ArrowUp":
        case "ArrowDown":
            event.preventDefault();
            handleKeyboardFocusNavigation(event.key, event.target as HTMLElement, getLiElements());
            break;

        case " ":
        case "Spacebar":
            event.preventDefault();
            setActiveItem(item);
            break;
    }
}

// Unique id to connect aria-labelledby with readonly label
function getAriaLabelledbyId(item: T): string {
    return `${props.elementId}_${itemKey(item)}`;
}

// Unique id to connect aria-labelledby with readonly label
function getItemId(item: T): string {
    return `${props.elementId}_item_${itemKey(item)}`;
}

// Focus effect is done with box-shadow.
// By setting position to relative the
// item and box-shadow is drawn with a higher z-index,
// thus no focus border under other list items.
function onItemFocus(event: FocusEvent): void {
    if (event && event.target) {
        (event.target as HTMLElement).style.position = "relative";
    }
}

function onItemBlur(event: FocusEvent): void {
    if (event && event.target) {
        (event.target as HTMLElement).style.position = "static";
    }
}

function callbackAfterItemAdd(item: T): void {
    setActiveItem(item);
}

function callbackBeforeItemDelete(item: T): void {
    if (internalItems.value.length === 0) {
        return;
    }
    let targetIndex = internalItems.value.indexOf(item) - 1;
    if (targetIndex < 0 && internalItems.value.length > 1) {
        targetIndex = 1;
    } else if (targetIndex < 0) {
        targetIndex = 0;
    }
    setActiveItem(internalItems.value[targetIndex]);
    const targetElement = getLiElements()[targetIndex];
    if (targetElement) {
        targetElement.focus();
    }
}

function isActive(item: T): boolean {
    return props.checkbox && itemEquals(activeItem.value, item, internalKey);
}
</script>

<template>
    <ul v-if="!selectable" class="list">
        <li v-for="item in internalItems" :key="itemKey(item)" class="list__item">
            <div ref="listItemPanes" class="list__item__itempane">
                <!--
                     @slot Slot for displaying an item. The item object is available through `v-slot="{ <propertyName> }"`, e.g. `v-slot="{ item }"`.
                     @binding {T} item - The object to be visualized.
                -->
                <slot v-bind="{ item }" />
            </div>
        </li>
        <li v-if="isEmpty" class="list__item">
            <div class="list__item__itempane">
                <!--
                     @slot Slot for displaying a message when list is empty. Default text is 'Listan är tom' (key fkui.list.empty).
                -->
                <slot name="empty">
                    <em>{{ $t("fkui.list.empty", "Listan är tom") }}</em>
                </slot>
            </div>
        </li>
    </ul>
    <ul v-else ref="ulElement" class="list list--hover">
        <li
            v-for="(item, index) in internalItems"
            :id="getItemId(item)"
            :key="itemKey(item)"
            :aria-labelledby="getItemId(item)"
            :class="itemClasses(item)"
            class="list__item"
            :tabindex="checkbox ? 0 : undefined"
            @keydown.self="onItemKeyDown($event, item)"
            @focus.self="onItemFocus"
            @blur.self="onItemBlur"
        >
            <div v-if="checkbox" class="list__item__selectpane" @click.self="onSelect(item)">
                <div class="list__item__selectpane__input">
                    <f-checkbox-field :value="true" :model-value="isSelected(item)" @click.self="onSelect(item)">
                        <span :id="getAriaLabelledbyId(item)" class="sr-only">
                            <!--
                                 @slot Slot for screen reader text when checkbox (to selecte item) get focus. The item object is available through `v-slot="{ <propertyName> }"`, e.g. `v-slot="{ item }"`.
                                 @binding {T} item - The object to be screen read.
                            -->
                            <slot name="screenreader" v-bind="{ item }" />
                        </span>
                    </f-checkbox-field>
                </div>
            </div>
            <component
                :is="checkbox ? 'div' : 'a'"
                ref="listItemPanes"
                :href="!checkbox ? 'javascript:' : undefined"
                class="list__item__itempane"
                @click="onItemClick($event, index, item)"
            >
                <slot v-bind="{ item }" />
            </component>
        </li>
        <li v-if="isEmpty" class="list__item">
            <div class="list__item__itempane">
                <slot name="empty">
                    <em>{{ $t("fkui.list.empty", "Listan är tom") }}</em>
                </slot>
            </div>
        </li>
    </ul>
</template>
