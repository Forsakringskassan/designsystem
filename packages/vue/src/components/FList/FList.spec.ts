import "html-validate/jest";
import { VueWrapper, mount } from "@vue/test-utils";
import { defineComponent } from "vue";
import { ListArray, UnknownItem } from "../../types";
import * as ListUtils from "../../utils/ListUtils";
import { TranslationPlugin } from "../../plugins";
import FList from "./FList.vue";

const items = [
    { id: "1", name: "TEST 1", permission: "Read" },
    { id: "2", name: "TEST 2", permission: "Write" },
    { id: "3", name: "TEST 3", permission: "Read" },
];

const oneItem = [{ id: "1", name: "TEST 1", permission: "Read" }];

function createWrapper({
    props = {},
    slots = {},
    screenReaderSlot = true,
    stubs = [] as string[],
}): VueWrapper {
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any -- technical debt
    const preDefinedSlots: any = {
        default: `<p>{{ params.item.id }}, {{ params.item.name }}, {{ params.item.permission }}</p>`,
        screenreader: /* HTML */ `<span>{{ params.item.name }}</span>`,
    };

    if (!screenReaderSlot) {
        delete preDefinedSlots.screenreader;
    }

    return mount(FList, {
        props: {
            items: [],
            keyAttribute: "id",
            ...props,
        },
        slots: { ...preDefinedSlots, ...slots },
        global: {
            plugins: [TranslationPlugin],
            stubs: stubs,
        },
    });
}

afterEach(() => {
    jest.restoreAllMocks();
});

describe("snapshots", () => {
    it("should match snapshot with a list containg three items", () => {
        const wrapper = createWrapper({
            props: {
                items,
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with list being selectable with second item active", async () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
            },
            stubs: ["f-checkbox-field"],
        });

        await wrapper
            .findAll("li")[1]
            .findAll(".list__item__itempane")[0]
            .trigger("click");

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with list being selectable with first and second item selected", () => {
        const preSelectedItems = [items[0], items[1]];
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
                modelValue: preSelectedItems,
            },
            stubs: ["f-checkbox-field"],
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with no items in list and default text", () => {
        const wrapper = createWrapper({
            props: {
                items: [],
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("should match snapshot with no items in list and custom text", () => {
        const wrapper = createWrapper({
            props: {
                items: [],
            },
            slots: {
                empty: "THE LIST IS EMPTY!",
            },
        });

        expect(wrapper.element).toMatchSnapshot();
    });
});

describe("active element", () => {
    it("should activate element and emit change event when clicking item", async () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
            },
        });
        const item = wrapper.findAll(".list__item__itempane")[1];
        await item.trigger("click");
        const li = item.element.closest("li")!;
        expect(Array.from(li.classList.values())).toMatchInlineSnapshot(`
            [
              "list__item--active",
              "list__item",
            ]
        `);

        expect(wrapper.emitted("change")![0][0]).toMatchInlineSnapshot(`
            {
              "id": "2",
              "name": "TEST 2",
              "permission": "Write",
            }
        `);
    });

    it.each([" ", "Spacebar"])(
        "should activate element and emit change event when item getting space key (%s) down event",
        async (key: string) => {
            const wrapper = createWrapper({
                props: {
                    items,
                    selectable: true,
                },
            });
            const li = wrapper.findAll("li")[1];
            await li.trigger("keydown", { key });

            expect(Array.from(li.element.classList.values())).toMatchSnapshot();
            expect(wrapper.emitted("change")![0][0]).toMatchSnapshot();
        },
    );

    it("should activate element and emit click event when clicking item", async () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
            },
        });
        const item = wrapper.findAll(".list__item__itempane")[1];
        await item.trigger("click");
        const li = item.element.closest("li")!;
        expect(Array.from(li.classList.values())).toMatchInlineSnapshot(`
            [
              "list__item--active",
              "list__item",
            ]
        `);

        expect(wrapper.emitted("click")![0][0]).toMatchInlineSnapshot(`
            {
              "id": "2",
              "name": "TEST 2",
              "permission": "Write",
            }
        `);
        expect(wrapper.emitted("click")).toHaveLength(1);
    });

    it.each([" ", "Spacebar"])(
        "should activate element and emit click event when item getting space key (%s) down event",
        async (key: string) => {
            const wrapper = createWrapper({
                props: {
                    items,
                    selectable: true,
                },
            });
            const li = wrapper.findAll("li")[1];
            await li.trigger("keydown", { key });

            expect(Array.from(li.element.classList.values())).toMatchSnapshot();
            expect(wrapper.emitted("click")![0][0]).toMatchSnapshot();
            expect(wrapper.emitted("click")!).toHaveLength(1);
        },
    );

    it("should emit click event when clicking on active item", async () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
            },
        });
        const item = wrapper.findAll(".list__item__itempane")[1];
        await item.trigger("click");
        await item.trigger("click");
        expect(wrapper.emitted("click")!).toHaveLength(2);
    });

    it.each([" ", "Spacebar"])(
        "should emit click event when active item getting space key (%s) down event",
        async (key: string) => {
            const wrapper = createWrapper({
                props: {
                    items,
                    selectable: true,
                },
            });
            const li = wrapper.findAll("li")[1];
            await li.trigger("keydown", { key });
            await li.trigger("keydown", { key });
            expect(wrapper.emitted("click")!).toHaveLength(2);
        },
    );

    it("should emit update:active when clicking on item", async () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
            },
        });
        const item = wrapper.findAll(".list__item__itempane")[1];
        await item.trigger("click");
        expect(wrapper.emitted("update:active")![0][0]).toMatchInlineSnapshot(`
            {
              "id": "2",
              "name": "TEST 2",
              "permission": "Write",
            }
        `);
    });

    it("should reset active item", async () => {
        let activeItem: UnknownItem | undefined;
        const TestComponent = defineComponent({
            name: "TestComponent",
            components: { FList },
            data() {
                return { items, activeItem, selectedItems: [] };
            },
            template: /* HTML */ `
                <f-list
                    v-model="selectedItems"
                    v-model:active="activeItem"
                    :items="items"
                    key-attribute="id"
                    selectable
                >
                    <template #default="{ item }">
                        <h3>{{ item.frukt }}</h3>
                    </template>
                    <template #screenreader="{ item }">
                        Frukt {{ item.frukt }}
                    </template>
                </f-list>
            `,
        });
        const wrapper = mount(TestComponent);
        wrapper.vm.$data.activeItem = items[1];
        await wrapper.vm.$nextTick();
        const listItem = wrapper.findAll(".list__item")[1];
        expect(listItem.classes()).toContain("list__item--active");
        wrapper.vm.$data.activeItem = undefined;
        await wrapper.vm.$nextTick();
        expect(listItem.classes()).not.toContain("list__item--active");
    });
});

describe("select events", () => {
    it("should emit select event when items selectpane is clicked", async () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
            },
        });

        await wrapper.findAll("div.list__item__selectpane")[1].trigger("click");
        expect(wrapper.emitted()["select"]).toEqual([[items[1]]]);
    });

    it("should emit select event when items checkbox is clicked", async () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
                keyAttribute: "id",
            },
        });

        await wrapper.findAll("input")[1].trigger("click");
        expect(wrapper.emitted()["select"]).toEqual([[items[1]]]);
    });

    it("should emit unselect event when items checkbox is clicked", async () => {
        const preSelected = [items[1]];
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
                modelValue: preSelected,
            },
        });

        await wrapper.findAll("input")[1].trigger("click");
        expect(wrapper.emitted()["unselect"]).toEqual([[items[1]]]);
    });
});

describe("v-model (update event)", () => {
    it("should update v-model when selecting and unselecting same item", async () => {
        const modelValue: ListArray = [];
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
                modelValue,
            },
        });
        const secondCheckbox = wrapper.findAll("input")[1];

        await secondCheckbox.trigger("click");
        expect(wrapper.emitted("update:modelValue")![0][0])
            .toMatchInlineSnapshot(`
            [
              {
                "id": "2",
                "name": "TEST 2",
                "permission": "Write",
              },
            ]
        `);

        await secondCheckbox.trigger("click");
        expect(
            wrapper.emitted("update:modelValue")![1][0],
        ).toMatchInlineSnapshot(`[]`);
    });

    it("should update v-model when selecting two different items", async () => {
        const modelValue: ListArray = [];
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
                modelValue,
            },
        });
        const firstCheckbox = wrapper.findAll("input")[0];
        const secondCheckbox = wrapper.findAll("input")[1];

        await firstCheckbox.trigger("click");
        expect(wrapper.emitted("update:modelValue")![0][0]).toEqual([
            {
                id: "1",
                name: "TEST 1",
                permission: "Read",
            },
        ]);

        await secondCheckbox.trigger("click");
        expect(wrapper.emitted("update:modelValue")![1][0]).toEqual([
            {
                id: "1",
                name: "TEST 1",
                permission: "Read",
            },
            {
                id: "2",
                name: "TEST 2",
                permission: "Write",
            },
        ]);
    });

    it("should update v-model when selecting and unselecting two items", async () => {
        const modelValue: ListArray = [];
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
                modelValue,
            },
        });
        const firstCheckbox = wrapper.findAll("input")[0];
        const secondCheckbox = wrapper.findAll("input")[1];

        await firstCheckbox.trigger("click");
        await secondCheckbox.trigger("click");

        expect(wrapper.emitted("update:modelValue")![0][0]).toEqual([
            {
                id: "1",
                name: "TEST 1",
                permission: "Read",
            },
            {
                id: "2",
                name: "TEST 2",
                permission: "Write",
            },
        ]);

        await firstCheckbox.trigger("click");
        await secondCheckbox.trigger("click");

        expect(
            wrapper.emitted("update:modelValue")![3][0],
        ).toMatchInlineSnapshot(`[]`);
    });

    it("should select items when updating v-model", async () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
            },
        });

        let allInputs = wrapper.element.querySelectorAll("input");
        expect(allInputs[0].checked).toBeFalsy();
        expect(allInputs[1].checked).toBeFalsy();
        expect(allInputs[2].checked).toBeFalsy();

        await wrapper.setProps({ modelValue: [items[0], items[2]] });
        await wrapper.vm.$nextTick();

        allInputs = wrapper.element.querySelectorAll("input");
        expect(allInputs[0].checked).toBeTruthy();
        expect(allInputs[1].checked).toBeFalsy();
        expect(allInputs[2].checked).toBeTruthy();
    });

    it("should update activeItem from v-model if provided or changed", async () => {
        const active = items[1];
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
                active,
            },
        });

        const liItems = wrapper.findAll("li");
        expect(Array.from(liItems[1].element.classList.values())).toContain(
            "list__item--active",
        );

        wrapper.setProps({ active: items[2] });
        await wrapper.vm.$nextTick();

        expect(Array.from(liItems[2].element.classList.values())).toContain(
            "list__item--active",
        );
    });

    it("should be able to preselect items", () => {
        const wrapper = createWrapper({
            props: {
                items,
                selectable: true,
                modelValue: [items[0], items[2]],
            },
        });

        const allInputs = wrapper.element.querySelectorAll("input");
        expect(allInputs[0].checked).toBeTruthy();
        expect(allInputs[1].checked).toBeFalsy();
        expect(allInputs[2].checked).toBeTruthy();
    });
});

describe("`keyAttribute`", () => {
    it("should not throw if valid and unique", async () => {
        expect.assertions(1);

        expect(() => {
            mount(FList, {
                props: {
                    keyAttribute: "id",
                    items: [{ id: "a" }, { id: "b" }, { id: "c" }],
                },
            });
        }).not.toThrow();
    });

    it("should throw error if not unique in items", async () => {
        expect.assertions(1);

        expect(() => {
            mount(FList, {
                props: {
                    keyAttribute: "id",
                    items: [{ id: "a" }, { id: "b" }, { id: "b" }],
                },
            });
        }).toThrowErrorMatchingInlineSnapshot(
            `"Expected each item to have key [id] with unique value but encountered duplicate of "b" in item index 2."`,
        );
    });

    it("should be optional", async () => {
        expect.assertions(1);

        expect(() => {
            mount(FList, {
                props: {
                    items: [{ id: "a" }, { id: "b" }, { id: "c" }],
                },
            });
        }).not.toThrow();
    });
});

describe("keyboard navigation", () => {
    it.each(["Up", "Down", "ArrowUp", "ArrowDown"])(
        "should pass key '%s' to handleKeyboardFocusNavigation method",
        async (key: string) => {
            const methodSpy = jest.spyOn(
                ListUtils,
                "handleKeyboardFocusNavigation",
            );

            const wrapper = createWrapper({
                props: {
                    items: oneItem,
                    selectable: true,
                },
            });
            // Need to use setProps to trigger Updated() on the component
            await wrapper.setProps({ items: items });
            const li = wrapper.findAll("li")[1];
            await li.trigger("keydown", { key });

            expect(methodSpy).toHaveBeenCalledWith(
                key,
                expect.anything(),
                expect.anything(),
            );
        },
    );
});

describe("screenreader slot", () => {
    /* eslint-disable-next-line no-console -- technical debt, bad practice and
     * the mock is not restored so leaking to other tests */
    console.error = jest.fn();
    const modelValue: ListArray = [];

    it("should throw exception if missing screenreader slot and selectable option", async () => {
        expect(() => {
            createWrapper({
                props: {
                    items,
                    selectable: true,
                    modelValue,
                },
                screenReaderSlot: false,
            });
        }).toThrowErrorMatchingInlineSnapshot(
            `"Slot "screenreader" is required when having "selectable" & "checkbox" option."`,
        );
    });
});

describe("html-validate", () => {
    it("should require `key-attribute` to be non-empty if used", () => {
        expect.assertions(1);
        expect('<f-list key-attribute=""></f-list>').not.toHTMLValidate({
            message: 'Attribute "key-attribute" has invalid value ""',
        });
    });

    it("should require items attribute", () => {
        expect.assertions(1);
        expect("<f-list></f-list>").not.toHTMLValidate({
            message: '<f-list> is missing required "items" attribute',
        });
    });
});
