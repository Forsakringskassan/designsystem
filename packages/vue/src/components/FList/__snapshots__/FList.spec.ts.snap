// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`active element should activate element and emit change event when item getting space key ( ) down event 1`] = `
[
  "list__item--active",
  "list__item",
]
`;

exports[`active element should activate element and emit change event when item getting space key ( ) down event 2`] = `
{
  "id": "2",
  "name": "TEST 2",
  "permission": "Write",
}
`;

exports[`active element should activate element and emit change event when item getting space key (Spacebar) down event 1`] = `
[
  "list__item--active",
  "list__item",
]
`;

exports[`active element should activate element and emit change event when item getting space key (Spacebar) down event 2`] = `
{
  "id": "2",
  "name": "TEST 2",
  "permission": "Write",
}
`;

exports[`active element should activate element and emit click event when item getting space key ( ) down event 1`] = `
[
  "list__item--active",
  "list__item",
]
`;

exports[`active element should activate element and emit click event when item getting space key ( ) down event 2`] = `
{
  "id": "2",
  "name": "TEST 2",
  "permission": "Write",
}
`;

exports[`active element should activate element and emit click event when item getting space key (Spacebar) down event 1`] = `
[
  "list__item--active",
  "list__item",
]
`;

exports[`active element should activate element and emit click event when item getting space key (Spacebar) down event 2`] = `
{
  "id": "2",
  "name": "TEST 2",
  "permission": "Write",
}
`;

exports[`snapshots should match snapshot with a list containg three items 1`] = `
<ul
  class="list"
>
  
  <li
    class="list__item"
  >
    <div
      class="list__item__itempane"
    >
      
      <p>
        1, TEST 1, Read
      </p>
      
    </div>
  </li>
  <li
    class="list__item"
  >
    <div
      class="list__item__itempane"
    >
      
      <p>
        2, TEST 2, Write
      </p>
      
    </div>
  </li>
  <li
    class="list__item"
  >
    <div
      class="list__item__itempane"
    >
      
      <p>
        3, TEST 3, Read
      </p>
      
    </div>
  </li>
  
  <!--v-if-->
</ul>
`;

exports[`snapshots should match snapshot with list being selectable with first and second item selected 1`] = `
<ul
  class="list list--hover"
>
  
  <li
    aria-labelledby="fkui-vue-element-0001_item_1"
    class="list__item--selected list__item"
    id="fkui-vue-element-0001_item_1"
    tabindex="0"
  >
    <div
      class="list__item__selectpane"
    >
      <div
        class="list__item__selectpane__input"
      >
        <f-checkbox-field-stub
          disabled="false"
          id="fkui-vue-element-0002"
          modelvalue="true"
          value="true"
        />
      </div>
    </div>
    <div
      class="list__item__itempane"
    >
      
      <p>
        1, TEST 1, Read
      </p>
      
    </div>
  </li>
  <li
    aria-labelledby="fkui-vue-element-0001_item_2"
    class="list__item--selected list__item"
    id="fkui-vue-element-0001_item_2"
    tabindex="0"
  >
    <div
      class="list__item__selectpane"
    >
      <div
        class="list__item__selectpane__input"
      >
        <f-checkbox-field-stub
          disabled="false"
          id="fkui-vue-element-0003"
          modelvalue="true"
          value="true"
        />
      </div>
    </div>
    <div
      class="list__item__itempane"
    >
      
      <p>
        2, TEST 2, Write
      </p>
      
    </div>
  </li>
  <li
    aria-labelledby="fkui-vue-element-0001_item_3"
    class="list__item"
    id="fkui-vue-element-0001_item_3"
    tabindex="0"
  >
    <div
      class="list__item__selectpane"
    >
      <div
        class="list__item__selectpane__input"
      >
        <f-checkbox-field-stub
          disabled="false"
          id="fkui-vue-element-0004"
          modelvalue="false"
          value="true"
        />
      </div>
    </div>
    <div
      class="list__item__itempane"
    >
      
      <p>
        3, TEST 3, Read
      </p>
      
    </div>
  </li>
  
  <!--v-if-->
</ul>
`;

exports[`snapshots should match snapshot with list being selectable with second item active 1`] = `
<ul
  class="list list--hover"
>
  
  <li
    aria-labelledby="fkui-vue-element-0001_item_1"
    class="list__item"
    id="fkui-vue-element-0001_item_1"
    tabindex="0"
  >
    <div
      class="list__item__selectpane"
    >
      <div
        class="list__item__selectpane__input"
      >
        <f-checkbox-field-stub
          disabled="false"
          id="fkui-vue-element-0002"
          modelvalue="false"
          value="true"
        />
      </div>
    </div>
    <div
      class="list__item__itempane"
    >
      
      <p>
        1, TEST 1, Read
      </p>
      
    </div>
  </li>
  <li
    aria-labelledby="fkui-vue-element-0001_item_2"
    class="list__item--active list__item"
    id="fkui-vue-element-0001_item_2"
    tabindex="0"
  >
    <div
      class="list__item__selectpane"
    >
      <div
        class="list__item__selectpane__input"
      >
        <f-checkbox-field-stub
          disabled="false"
          id="fkui-vue-element-0003"
          modelvalue="false"
          value="true"
        />
      </div>
    </div>
    <div
      class="list__item__itempane"
    >
      
      <p>
        2, TEST 2, Write
      </p>
      
    </div>
  </li>
  <li
    aria-labelledby="fkui-vue-element-0001_item_3"
    class="list__item"
    id="fkui-vue-element-0001_item_3"
    tabindex="0"
  >
    <div
      class="list__item__selectpane"
    >
      <div
        class="list__item__selectpane__input"
      >
        <f-checkbox-field-stub
          disabled="false"
          id="fkui-vue-element-0004"
          modelvalue="false"
          value="true"
        />
      </div>
    </div>
    <div
      class="list__item__itempane"
    >
      
      <p>
        3, TEST 3, Read
      </p>
      
    </div>
  </li>
  
  <!--v-if-->
</ul>
`;

exports[`snapshots should match snapshot with no items in list and custom text 1`] = `
<ul
  class="list"
>
  
  
  <li
    class="list__item"
  >
    <div
      class="list__item__itempane"
    >
      
      THE LIST IS EMPTY!
      
    </div>
  </li>
</ul>
`;

exports[`snapshots should match snapshot with no items in list and default text 1`] = `
<ul
  class="list"
>
  
  
  <li
    class="list__item"
  >
    <div
      class="list__item__itempane"
    >
      
      <em>
        Listan Ã¤r tom
      </em>
      
    </div>
  </li>
</ul>
`;
